framework/Auth/lib/Horde/Auth/Fallback.php

```diff
@@ -0,0 +1,229 @@
+<?php
+/**
+ * Copyright 2016-2017 Horde LLC (http://www.horde.org/)
+ *
+ * See the enclosed file COPYING for license information (LGPL). If you did
+ * not receive this file, http://www.horde.org/licenses/lgpl21
+ *
+ * @author   Ralf Lang <lang@b1-systems.de>
+ * @category Horde
+ * @license http://www.horde.org/licenses/lgpl21 LGPL-2.1
+ * @package  Auth
+ */
+/**
+ * The Horde_Auth_Fallback class provides a way to combine two separate
+ * drivers for failover and legacy support cases, for example
+ * using ldap, but falling back to a local database
+ * To the user, this driver presents the combined, unique users of both backends
+ * Only the primary driver allows adding, editing and removing users.
+ *
+ * @author   Ralf Lang <lang@b1-systems.de>
+ * @category Horde
+ * @license http://www.horde.org/licenses/lgpl21 LGPL-2.1
+ * @package  Auth
+ */
+class Horde_Auth_Fallback extends Horde_Auth_Base
+{
+    /**
+     * Constructor.
+     *
+     * @param array $params  Required parameters:
+     * <pre>
+     * 'primary_driver' - (Horde_Auth_Base) The primary driver.
+     * 'fallback_driver' - (Horde_Auth_Base) The auth driver.
+     * 'capabilities' -  (array) Keys each 'adduser', 'removeuser', 'updateuser', 'listusers', 'resetpassword'
+     *                   Values "both", "primary", "fallback", "none"
+     *                   See method doc for "both" case on resetpassword and listusers
+     *                   Also overrides capabilities exposed
+     * </pre>
+     *
+     * @throws InvalidArgumentException
+     */
+    public function __construct(array $params = array())
+    {
+        foreach (array('primary_driver', 'fallback_driver', 'capabilities') as $val) {
+            if (!isset($params[$val])) {
+                throw new InvalidArgumentException('Missing ' . $val . ' parameter.');
+            }
+        }
+        parent::__construct($params);
+    }
+    /**
+     * Find out if a set of login credentials are valid.
+     *
+     * @param string $userId      The userId to check.
+     * @param array $credentials  The credentials to use.
+     *
+     * @throws Horde_Auth_Exception
+     */
+    protected function _authenticate($userId, $credentials)
+    {
+        if (!$this->_params['primary_driver']->authenticate($userId, $credentials) && !$this->_params['fallback_driver']->authenticate($userId, $credentials)) {
+            throw new Horde_Auth_Exception($this->_params['primary_driver']->getError(true), $this->_params['primary_driver']->getError());
+        }
+    }
+    /**
+     * Trust user-supplied capabilities for writing operations. Else,
+     * Query the primary Auth object to find out if it supports the given
+     * capability.
+     *
+     * @param string $capability  The capability to test for.
+     *
+     * @return boolean  Whether or not the capability is supported.
+     */
+    public function hasCapability($capability)
+    {
+        $lcCapability = strtolower($capability);
```

Comment:
But do you actually use this? If yes, it seems to be an exception (that I missed), so by default, you should still loop.